locator/token_metadata.hh

```diff
@@ -267,6 +268,16 @@ public:
     // returns empty vector if keyspace_name not found.
     inet_address_vector_topology_change pending_endpoints_for(const token& token, const sstring& keyspace_name) const;
 
+    // This function returns a list of nodes to which a read request should be directed.
+    // Returns not null only during topology changes, if _topology_change_stage == read_new and
+    // new set of replicas differs from the old one.
+    std::optional<inet_address_vector_replica_set> endpoints_for_reading(const token& token, const sstring& keyspace_name) const;
```

Comment:
I think it would be better to use it unconditionally. But okay to discuss it post-merge.

Reply:
This would mean complete duplication of the `replication_map` in vnode_erm. Here I tried to be memory efficient (store only ranges with diffs only during `write_both_read_new`) and do it similarly to `pending_ranges` for writes to reduce the scope of changes.

`replication_map` in `vnode_erm` seems more efficient than these interval trees - it's just a sorted vector + has_map. In [this follow-up PR](https://github.com/scylladb/scylladb/pull/13715) I move these interval trees into erm. We could possibly go further and turn them into hash_maps: vnode -> pending_endpoints/read_endpoints. We can't just merge them into natural_endpoints since each `natural_endpoints`, `pending_endpoints` and `pending_endpoints` have their own distinct usages.

Reply:
Yes. How about this:

1. Wait for @bhalevy to switch these to use node* (or uint16_t index to a node vector) instead of inet_address_vector_blah_blah_blah.
2. Store those ranges contiguously as sub-spans of a large vector (to avoid allocating an object per range)
3. change the APIs to return std::span\<node_index_t>

Reply:
https://github.com/scylladb/scylladb/issues/13839